src/hotspot/share/services/diagnosticCommand.cpp

```diff
@@ -1177,6 +1180,78 @@ void CompilationMemoryStatisticDCmd::execute(DCmdSource source, TRAPS) {
   CompilationMemoryStatistic::print_all_by_size(output(), human_readable, minsize);
 }
 
+VMDebugDCmd::VMDebugDCmd(outputStream* output, bool heap) :
+                                     DCmdWithParser(output, heap),
+  _subcommand("subcommand", "", "STRING", true, nullptr),
+  _arg1("arg1", "", "STRING", false, nullptr),
+  _arg2("arg2", "", "STRING", false, nullptr),
+  _arg3("arg3", "", "STRING", false, nullptr),
+  _verbose("-verbose", "", "BOOLEAN", false, "false") {
+
+  _dcmdparser.add_dcmd_argument(&_subcommand);
+  _dcmdparser.add_dcmd_argument(&_arg1);
+  _dcmdparser.add_dcmd_argument(&_arg2);
+  _dcmdparser.add_dcmd_argument(&_arg3);
+  _dcmdparser.add_dcmd_option(&_verbose);
+}
+
+void VMDebugDCmd::find() {
+  if (!_arg1.has_value()) {
+    output()->print_cr("missing argument");
+  } else {
+    intptr_t x = strtoll(_arg1.value(), nullptr, 0);
+    if (!os::is_readable_pointer((intptr_t*) x)) {
+      output()->print_cr("address not safe");
+    } else {
+      if (Universe::heap()->is_in((oopDesc*) x)) {
+        if (x != align_down(x, ObjectAlignmentInBytes)) {
+          output()->print_cr("misaligned oop");
+          return;
+        }
+        if (!dbg_is_good_oop_detailed((oopDesc*) x)) {
+          output()->print_cr("bad oop");
+          return;
+        }
+      }
+      os::print_location(output(), x, _verbose.is_set());
+    }
+  }
+}
+
+void VMDebugDCmd::execute(DCmdSource source, TRAPS) {
+  DebuggingContext dc{}; // avoid asserts
+
+  // Interpret _subcommand, using further args _arg1, etc as required.
+  if (strcmp("events", _subcommand.value()) == 0) {
+    Events::print_all(output(), 100);
+  } else if (strcmp("threads", _subcommand.value()) == 0) {
+    char buf[1024];
+    Threads::print_on_error(output(), THREAD, buf, sizeof(buf));
+  } else if (strcmp("findclass", _subcommand.value()) == 0) {
+    if (!_arg1.has_value() || !_arg2.has_value()) {
+      output()->print_cr("missing argument");
+    } else {
+      long flags = strtol(_arg2.value(), nullptr, 0);
+      ClassPrinter::print_classes(_arg1.value(), flags, output());
+    }
+  } else if (strcmp("findmethod", _subcommand.value()) == 0) {
+    if (!_arg1.has_value() || !_arg2.has_value() || !_arg3.has_value()) {
+      output()->print_cr("missing argument");
+    } else {
+      long flags = strtol(_arg3.value(), nullptr, 0);
+      ClassPrinter::print_methods(_arg1.value(), _arg2.value(), flags, output());
+    }
+  } else if (strcmp("find", _subcommand.value()) == 0) {
+    if (!UnlockDiagnosticVMOptions) {
```

Comment:
Would it make sense to require enabling `UnlockDiagnosticVMOptions` for all sub-commands, so that it is clear this is not for live production use?

Reply:
Yes sure I've made that apply to all of them for consistency.